src/mark.c

```diff
@@ -203,55 +206,87 @@ gboolean mark_run(const gchar *state,
 		gchar **slot_name,
 		gchar **message)
 {
-	RaucSlot *slot = NULL;
-	GError *ierror = NULL;
-	gboolean res;
+	g_autoptr(GList) slots = NULL;
+	g_autoptr(GError) ierror = NULL;
 
 	g_assert(slot_name == NULL || *slot_name == NULL);
 	g_assert(message != NULL && *message == NULL);
 
-	slot = get_slot_by_identifier(slot_identifier, &ierror);
+	slots = get_slots_by_identifier(slot_identifier, &ierror);
 	if (ierror) {
-		res = FALSE;
 		*message = g_strdup(ierror->message);
-		goto out;
+		return FALSE;
 	}
 
 	if (g_strcmp0(state, "good") == 0) {
+		if (g_list_length(slots) > 1) {
+			*message = g_strdup("cannot mark multiple slots as good");
+			if (slot_name)
+				*slot_name = g_strdup("other");
+			return FALSE;
+		}
+		RaucSlot *slot = (RaucSlot*)g_list_first(slots)->data;
+
 		if (!r_mark_good(slot, &ierror)) {
-			res = FALSE;
 			*message = g_strdup(ierror->message);
-		} else {
-			res = TRUE;
-			*message = g_strdup_printf("marked slot %s as good", slot->name);
+			return FALSE;
+		}
+
+		if (slot_name)
+			*slot_name = g_strdup(slot->name);
+		*message = g_strdup_printf("marked slot %s as good", slot->name);
```

Comment:
This string overwritten in the prevent-fallback case. Move it into an `else` branch below.